File: .vscode/launch_target.py
"""
This script is a shim that allows running bazel python targets by vscode debugger.
"""

import os
import runpy
import shlex
import subprocess
from pathlib import Path
from typing import cast

BAZEL_WORKSPACE = Path(__file__).resolve().parents[1]
BAZEL_BIN = "bazel"


def run_bazel_target(target: str) -> None:
    # Make sure the target is built.
    subprocess.check_call([BAZEL_BIN, "build", target], cwd=BAZEL_WORKSPACE)

    # What we could put together to get the binary and CWD it should run from.
    query = subprocess.check_output(
        [
            BAZEL_BIN,
            "cquery",
            "--ui_event_filters=-info,-debug",
            "--noshow_progress",
            "--output=starlark",
            "--starlark:expr=[target.files_to_run.executable.path, target.files_to_run.runfiles_manifest.path]",
            target,
        ],
        text=True,
        cwd=BAZEL_WORKSPACE,
    )

    execroot = subprocess.check_output([BAZEL_BIN, "info", "execution_root"], text=True, cwd=BAZEL_WORKSPACE)
    execroot = execroot.strip()

    
    # Not sure how to properly deal with this, but if the target is used as a tool, it might
    # return multiple values (on in out/...-fastabuild one in out/...opt-exec)
    # This is just a heuristic to get the last line
    last_query = [item for item in query.split("\n") if item][-1]

    binary, runfiles_manifest = cast(tuple[str, str], eval(last_query))

    binary_abs = BAZEL_WORKSPACE / binary
    runfiles_manifest_abs = BAZEL_WORKSPACE / runfiles_manifest

    # Bazel runs things in execroot.
    os.chdir(execroot)
    # Set the runfile manifest, in case the binary uses runfiles
    os.environ["RUNFILES_MANIFEST_FILE"] = str(runfiles_manifest_abs)
    # Run the binary!
    runpy.run_path(str(binary_abs), run_name="__main__")


if __name__ == "__main__":
    import sys

    # For vscode to be able to pass arguments in one go from command line, we need to
    # read them as one string :(
    if len(sys.argv) != 2:
        raise ValueError(
            'Expected only and only one argument, syntax: "<bazel target> <args>" as a single bash argument.'
        )

    args = shlex.split(sys.argv[1])

    target = args.pop(0)
    # Modify sys.argv, so when we transition to run the actual executable, it will have
    # all the arguments, minus the one that belongs to this script.
    sys.argv[1:] = args

    run_bazel_target(target)

File: .vscode/launch_target.go
// This is work in progress
// Does not work now.

package main

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
)

var (
	bazelWorkspace = getWorkspaceRoot()
	bazelBin       = "./bazel"
)

func main() {
	if len(os.Args) != 2 {
		fmt.Println("Expected exactly one argument: '<bazel target> <args>' as a single string.")
		os.Exit(1)
	}

	args := strings.Fields(os.Args[1])
	if len(args) == 0 {
		fmt.Println("No target specified.")
		os.Exit(1)
	}

	target := args[0]
	os.Args = append([]string{os.Args[0]}, args[1:]...)

	err := runBazelTarget(target)
	if err != nil {
		fmt.Printf("Failed to run bazel target: %v\n", err)
		os.Exit(1)
	}
}

func runBazelTarget(target string) error {
	// Build the target
	fmt.Println("Building target:", target)
	cmd := exec.Command(bazelBin, "build", target)
	cmd.Dir = bazelWorkspace
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr

	if err := cmd.Run(); err != nil {
		return fmt.Errorf("bazel build failed: %w", err)
	}

	// Query the target to get executable and runfiles
	queryCmd := exec.Command(
		bazelBin, "cquery",
		"--ui_event_filters=-info,-debug",
		"--noshow_progress",
		"--output=starlark",
		"--starlark:expr=[target.files_to_run.executable.path, target.files_to_run.runfiles_manifest.path]",
		target,
	)
	queryCmd.Dir = bazelWorkspace

	var out bytes.Buffer
	queryCmd.Stdout = &out
	queryCmd.Stderr = os.Stderr

	if err := queryCmd.Run(); err != nil {
		return fmt.Errorf("bazel cquery failed: %w", err)
	}

	// Get the execution root
	execRootCmd := exec.Command(bazelBin, "info", "execution_root")
	execRootCmd.Dir = bazelWorkspace
	execRoot, err := execRootCmd.Output()
	if err != nil {
		return fmt.Errorf("failed to get execroot: %w", err)
	}

	execRootStr := strings.TrimSpace(string(execRoot))

	// Parse the last line of the query output
	queryOutput := strings.TrimSpace(out.String())
	lines := strings.Split(queryOutput, "\n")
	if len(lines) == 0 {
		return errors.New("no output from cquery")
	}
	lastLine := lines[len(lines)-1]

	// Extract binary and runfiles manifest
	var binary, runfilesManifest string
	var result []string
	err = json.Unmarshal([]byte(lastLine), &result)
	if err != nil {
		return fmt.Errorf("failed to parse cquery output: %w", err)
	}
	if len(result) != 2 {
		return errors.New("unexpected cquery output format")
	}
	binary, runfilesManifest = result[0], result[1]

	binary = strings.Trim(binary, "\"")
	runfilesManifest = strings.Trim(runfilesManifest, "\"")

	binaryAbs := filepath.Join(bazelWorkspace, binary)
	runfilesManifestAbs := filepath.Join(bazelWorkspace, runfilesManifest)

	// Set environment variables and change directory to execroot
	os.Setenv("RUNFILES_MANIFEST_FILE", runfilesManifestAbs)
	err = os.Chdir(execRootStr)
	if err != nil {
		return fmt.Errorf("failed to change directory to execroot: %w", err)
	}

	// Execute the binary
	fmt.Println("Running binary:", binaryAbs)
	execCmd := exec.Command(binaryAbs, os.Args[1:]...)
	execCmd.Stdout = os.Stdout
	execCmd.Stderr = os.Stderr

	return execCmd.Run()
}

// getWorkspaceRoot finds the Bazel workspace root by walking up the directory tree.
func getWorkspaceRoot() string {
	dir, err := os.Getwd()
	if err != nil {
		fmt.Println("Failed to get current directory:", err)
		os.Exit(1)
	}

	for {
		if _, err := os.Stat(filepath.Join(dir, "MODULE.bazel")); err == nil {
			return dir
		}
		parent := filepath.Dir(dir)
		if parent == dir {
			break
		}
		dir = parent
	}

	fmt.Println("Bazel WORKSPACE file not found.")
	os.Exit(1)
	return ""
}
